File: Betting app/PushManager.swift
//
//  PushManager.swift
//  Betting app
//
//  Captures the APNs device token and persists it (per signed-in user) into
//  the `device_tokens` table, so the `send-push` Edge Function can deliver
//  per-game notifications (new picks, score changes, results).
//
//  Flow:
//    • registerIfAuthorized() — call on launch / after the onboarding
//      permission prompt. If notifications are authorized it asks iOS for a
//      token via registerForRemoteNotifications().
//    • AppDelegate.didRegisterForRemoteNotificationsWithDeviceToken → setDeviceToken
//    • upload() persists the token whenever we have BOTH a token and a session
//      (the token can arrive before sign-in, or the user can sign in later).
//
import UIKit
import Supabase

@MainActor
final class PushManager {
    static let shared = PushManager()
    private init() {}

    private var pendingToken: String?

    /// Ask iOS for an APNs token, but only if the user has granted permission.
    func registerIfAuthorized() {
        UNUserNotificationCenter.current().getNotificationSettings { settings in
            switch settings.authorizationStatus {
            case .authorized, .provisional, .ephemeral:
                DispatchQueue.main.async {
                    UIApplication.shared.registerForRemoteNotifications()
                }
            default:
                break
            }
        }
    }

    /// Called by AppDelegate when APNs returns the device token.
    func setDeviceToken(_ data: Data) {
        pendingToken = data.map { String(format: "%02x", $0) }.joined()
        Task { await upload() }
    }

    /// Best-effort: re-attempt the upload (e.g. after the user signs in, when
    /// a token was already captured during onboarding).
    func uploadIfPending() {
        guard pendingToken != nil else { return }
        Task { await upload() }
    }

    private struct TokenRow: Encodable {
        let token: String
        let user_id: String
        let environment: String
        let platform: String
        let app_version: String?
        /// The user's chosen app language ("en"/"fr"/"es"/…), so the
        /// send-push function can localize every notification server-side.
        let locale: String
    }

    private func upload() async {
        guard let token = pendingToken,
              let userId = SupabaseManager.client.auth.currentSession?.user.id
        else { return }
        #if DEBUG
        let environment = "sandbox"
        #else
        let environment = "production"
        #endif
        let version = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
        do {
            try await SupabaseManager.client
                .from("device_tokens")
                .upsert(
                    TokenRow(token: token,
                             user_id: userId.uuidString,
                             environment: environment,
                             platform: "ios",
                             app_version: version,
                             locale: LocalizationManager.shared.languageCode),
                    onConflict: "token"
                )
                .execute()
        } catch {
            // Best-effort — a failed token upload must never disrupt the app.
        }
    }
}

File: android/app/src/main/java/com/pick1/app/push/PushManager.kt
package com.pick1.app.push

import com.google.firebase.messaging.FirebaseMessaging
import com.pick1.app.BuildConfig
import com.pick1.app.data.LanguageManager
import com.pick1.app.data.Supabase
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.from
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable

/**
 * Registers this device's FCM token in the SAME `device_tokens` table the
 * iOS app writes to — the table already carries a `platform` column, so
 * Android rows sit alongside the iOS APNs rows and the `send-push` Edge
 * Function fans out to both once it grows an FCM branch.
 *
 * Mirrors `PushManager.swift`: upsert on token, stamp environment /
 * platform / app_version / locale so pushes can be localized server-side.
 */
object PushManager {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    @Serializable
    private data class TokenRow(
        val token: String,
        val user_id: String,
        val environment: String,
        val platform: String,
        val app_version: String?,
        val locale: String,
    )

    /** Fetch the current FCM token and upload it (call after sign-in). */
    fun register() {
        FirebaseMessaging.getInstance().token.addOnSuccessListener(::upload)
    }

    fun upload(token: String) {
        val userId = Supabase.client.auth.currentUserOrNull()?.id ?: return
        scope.launch {
            // Best-effort — a failed token upload must never disrupt the app.
            runCatching {
                Supabase.client.from("device_tokens").upsert(
                    TokenRow(
                        token = token,
                        user_id = userId,
                        environment = if (BuildConfig.DEBUG) "sandbox" else "production",
                        platform = "android",
                        app_version = BuildConfig.VERSION_NAME,
                        locale = LanguageManager.current ?: "en",
                    )
                ) { onConflict = "token" }
            }
        }
    }
}
